<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GeneTrajectory Tutorial • GeneTrajectory</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GeneTrajectory Tutorial">
<meta property="og:description" content="GeneTrajectory">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GeneTrajectory</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/GeneTrajectory.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GeneTrajectory Tutorial</h1>
                        <h4 data-toc-skip class="author">Rihao Qu,
Francesco Strino</h4>
            
            <h4 data-toc-skip class="date">03/20/2024</h4>
      
      
      <div class="hidden name"><code>GeneTrajectory.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>GeneTrajectory is a method for inferring gene trajectories in
scRNA-seq data, which facilitates understanding of gene dynamics
underlying biological processes. The major workflow of GeneTrajectory
comprises the following four main steps:</p>
<ul>
<li><p>Step 1. Build a cell-cell kNN graph in which each cell is
connected to its k-nearest neighbors. Find the shortest path connecting
each pair of cells in the graph and denote its length as the graph
distance between cells.</p></li>
<li><p>Step 2. Compute pairwise graph-based Wasserstein distance between
gene distributions, which quantifies the minimum cost of transporting
the distribution of a given gene into the distribution of another gene
in the cell graph.</p></li>
<li><p>Step 3. Generate a low-dimensional representation of genes (using
Diffusion Map by default) based on the gene-gene Wasserstein distance
matrix. Identify gene trajectories in a sequential manner.</p></li>
<li><p>Step 4. Determine the order of genes along each gene
trajectory.</p></li>
</ul>
<p><img src="GT_workflow.png"></p>
</div>
<div class="section level2">
<h2 id="preparation">Preparation<a class="anchor" aria-label="anchor" href="#preparation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##### Load required R libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">GeneTrajectory</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">plot3D</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-example-data">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data"></a>
</h2>
<p>The standard preprocessing can be done by employing the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Seurat</a>
R package which includes: library normalization; finding variable
features; scaling; generating PCA embedding (and UMAP embedding for
visualization).</p>
<p>The preprocessed Seurat object for this tutorial can be downloaded
from <a href="https://figshare.com/articles/dataset/Processed_Seurat_objects_for_GeneTrajectory_inference_Gene_Trajectory_Inference_for_Single-cell_Data_by_Optimal_Transport_Metrics_/25243225" class="external-link">figshare</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import the tutorial dataset and relabel the celltypes for visualization</span></span>
<span><span class="va">data_S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../../data/human_myeloid_seurat_obj.rds"</span><span class="op">)</span></span>
<span><span class="va">cluster_relabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span> <span class="op">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span>                     <span class="st">"1"</span> <span class="op">=</span> <span class="st">"Intermediate monocytes"</span>,</span>
<span>                     <span class="st">"2"</span> <span class="op">=</span> <span class="st">"CD16+ monocytes"</span>,</span>
<span>                     <span class="st">"3"</span> <span class="op">=</span> <span class="st">"Myeloid type-2 dendritic cells"</span><span class="op">)</span></span>
<span><span class="va">data_S</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">cluster_relabel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">data_S</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">data_S</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, shuffle <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneTrajectory_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="gene-gene-distance-computation">Gene-gene distance computation<a class="anchor" aria-label="anchor" href="#gene-gene-distance-computation"></a>
</h2>
<p>In this tutorial, we demonstrate gene-gene distance computation by
selecting the top 500 variable genes expressed by 1% to 50% of cells.
The expected running time for this computation is approximately 1 to 3
minutes.</p>
<div class="section level3">
<h3 id="select-genes">Select genes<a class="anchor" aria-label="anchor" href="#select-genes"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assay</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">data_S</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">assay</span></span>
<span><span class="va">data_S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">data_S</span>, nfeatures <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">all_genes</span> <span class="op">&lt;-</span> <span class="va">data_S</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="va">assay</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span></span>
<span><span class="va">expr_percent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">data_S</span><span class="op">[[</span><span class="va">assay</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">data</span><span class="op">[</span><span class="va">all_genes</span>, <span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data_S</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">all_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">expr_percent</span> <span class="op">&gt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">expr_percent</span> <span class="op">&lt;</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 251</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="prepare-the-input-for-gene-gene-distance-computation">Prepare the input for gene-gene distance computation<a class="anchor" aria-label="anchor" href="#prepare-the-input-for-gene-gene-distance-computation"></a>
</h3>
<p>Next, we compute the Diffusion Map cell embedding using
<code>RunDM</code>. We then use <code>GetGraphDistance</code> to
construct the cell-cell kNN graph (<code>K</code>=10 in this example)
and calculate cell-cell graph distances (defined by the length of the
shortest path connecting two cells in the kNN graph). To improve
computation efficiency, we coarse-grain the cell graph by grouping cells
into <code>N</code>=500 “meta-cells”.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the Diffusion Map cell embedding</span></span>
<span><span class="va">data_S</span> <span class="op">&lt;-</span> <span class="fu">GeneTrajectory</span><span class="fu">::</span><span class="fu"><a href="../reference/RunDM.html">RunDM</a></span><span class="op">(</span><span class="va">data_S</span><span class="op">)</span></span>
<span><span class="co"># Calculate cell-cell graph distances over a cell-cell kNN graph</span></span>
<span><span class="va">cell.graph.dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetGraphDistance.html">GetGraphDistance</a></span><span class="op">(</span><span class="va">data_S</span>, K <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Constructing kNN graph</span></span></code></pre>
<pre><code><span><span class="co">## Constructing graph distance matrix</span></span></code></pre>
<pre><code><span><span class="co">## The largest graph distance is 35</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Coarse-grain the cell graph by grouping cells into `N`=500 "meta-cells"</span></span>
<span><span class="va">cg_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoarseGrain.html">CoarseGrain</a></span><span class="op">(</span><span class="va">data_S</span>, <span class="va">cell.graph.dist</span>, <span class="va">genes</span>, N <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Run k-means clustering</span></span></code></pre>
<pre><code><span><span class="co">## Coarse-grain matrices</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="compute-gene-gene-distances">Compute gene-gene distances<a class="anchor" aria-label="anchor" href="#compute-gene-gene-distances"></a>
</h3>
<p>The computation of gene-gene Wasserstein distances is by employing
the package of Python Optimal Transport (<a href="https://pythonot.github.io/" class="external-link">POT</a>). Here, we recommend users to
set up a virtualenv using <a href="https://cran.r-project.org/web/packages/reticulate/index.html" class="external-link">reticulate</a>
R package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a virtualenv using reticulate</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/virtualenv-tools.html" class="external-link">virtualenv_exists</a></span><span class="op">(</span><span class="st">'gene_trajectory'</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/virtualenv-tools.html" class="external-link">virtualenv_create</a></span><span class="op">(</span><span class="st">'gene_trajectory'</span>, packages<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'gene_trajectory'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_virtualenv</a></span><span class="op">(</span><span class="st">'gene_trajectory'</span><span class="op">)</span></span>
<span><span class="co"># Import the function to compute gene-gene distances</span></span>
<span><span class="va">cal_ot_mat_from_numpy</span> <span class="op">&lt;-</span> <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/import.html" class="external-link">import</a></span><span class="op">(</span><span class="st">'gene_trajectory.compute_gene_distance_cmd'</span><span class="op">)</span><span class="op">$</span><span class="va">cal_ot_mat_from_numpy</span></span>
<span><span class="co"># Compute gene-gene distances (this is usually the most time-consuming step, which can take hours if the computation is done on thousands of genes over a large cell graph.)</span></span>
<span><span class="va">gene.dist.mat</span> <span class="op">&lt;-</span> <span class="fu">cal_ot_mat_from_numpy</span><span class="op">(</span>ot_cost <span class="op">=</span> <span class="va">cg_output</span><span class="op">[[</span><span class="st">"graph.dist"</span><span class="op">]</span><span class="op">]</span>, gene_expr <span class="op">=</span> <span class="va">cg_output</span><span class="op">[[</span><span class="st">"gene.expression"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene.dist.mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cg_output</span><span class="op">[[</span><span class="st">"features"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene.dist.mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cg_output</span><span class="op">[[</span><span class="st">"features"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">gene.dist.mat</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="gene-trajectory-inference-and-visualization">Gene trajectory inference and visualization<a class="anchor" aria-label="anchor" href="#gene-trajectory-inference-and-visualization"></a>
</h2>
<p>After we get the gene-gene distance matrix, we construct the gene
embedding by employing Diffusion Map.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_embedding</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetGeneEmbedding.html">GetGeneEmbedding</a></span><span class="op">(</span><span class="va">gene.dist.mat</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">$</span><span class="va">diffu.emb</span></span></code></pre></div>
<p>The extraction of gene trajectories is done sequentially. The initial
terminus node (terminus-1) is defined by the gene with the largest
distance from the origin in the Diffusion Map embedding. A random-walk
procedure is then employed on the gene graph to select the other genes
that belong to the trajectory terminated at terminus-1. After retrieving
genes for the first trajectory, we identify the terminus of the
subsequent gene trajectory among the remaining genes and repeat the
steps above. This is done sequentially until all detectable trajectories
are extracted.</p>
<ul>
<li><p><code>N</code>: the number of trajectories to extract;</p></li>
<li><p><code>t.list</code>: the time step of random walk to extract each
trajectory. Increasing <code>t</code> extracts a longer
trajectory.</p></li>
<li><p><code>K</code>: this should be consistent with the <code>K</code>
used in the previous step (<code>GetGeneEmbedding</code>).</p></li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract 3 gene trajectories</span></span>
<span><span class="va">gene_trajectory</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractGeneTrajectory.html">ExtractGeneTrajectory</a></span><span class="op">(</span><span class="va">gene_embedding</span>, <span class="va">gene.dist.mat</span>, N <span class="op">=</span> <span class="fl">3</span>, t.list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">7</span>,<span class="fl">7</span><span class="op">)</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">gene_trajectory</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Trajectory-1 Trajectory-2 Trajectory-3 </span></span>
<span><span class="co">##           43           55          153</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize gene trajectories</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/plot3D/man/scatter.html" class="external-link">scatter3D</a></span><span class="op">(</span><span class="va">gene_embedding</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>          <span class="va">gene_embedding</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>          <span class="va">gene_embedding</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>,</span>
<span>          bty <span class="op">=</span> <span class="st">"b2"</span>, colvar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">gene_trajectory</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span>,</span>
<span>          main <span class="op">=</span> <span class="st">"trajectory"</span>, pch <span class="op">=</span> <span class="fl">19</span>, cex <span class="op">=</span> <span class="fl">1</span>, theta <span class="op">=</span> <span class="fl">45</span>, phi <span class="op">=</span> <span class="fl">0</span>,</span>
<span>          col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plot3D/man/colors.html" class="external-link">ramp.col</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://scales.r-lib.org/reference/hue_pal.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneTrajectory_files/figure-html/unnamed-chunk-8-1.png" width="432"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the ordered list of genes along each gene trajectory</span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">gene_trajectory_sub</span> <span class="op">&lt;-</span> <span class="va">gene_trajectory</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">gene_trajectory</span><span class="op">$</span><span class="va">selected</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Trajectory-"</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_trajectory_sub</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">gene_trajectory_sub</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Pseudoorder"</span>, <span class="va">i</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">gene_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">genes</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-gene-bin-plots">Visualize gene bin plots<a class="anchor" aria-label="anchor" href="#visualize-gene-bin-plots"></a>
</h2>
<p>To examine how each given gene trajectory is reflected over the cell
graph, we can track how these genes are expressed across different
regions in the cell embedding. Here, we would recommend users to apply
<a href="https://github.com/KlugerLab/ALRA/blob/master/README.md" class="external-link">ALRA</a>
imputation to smooth the expression values for generating gene bin
plots.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SeuratWrappers</span><span class="op">)</span></span>
<span><span class="va">data_S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratWrappers/man/RunALRA.html" class="external-link">RunALRA</a></span><span class="op">(</span><span class="va">data_S</span><span class="op">)</span></span></code></pre></div>
<p>We define gene bin score by splitting each gene trajectory into 5
consecutive gene bins.</p>
<ul>
<li><p><code>N.bin</code>: the number of gene bins along each gene
trajectory;</p></li>
<li><p><code>trajectories</code>: gene trajectories to define gene bin
score;</p></li>
<li><p><code>assay</code>: assay used to compute the score of each gene
bin;</p></li>
<li><p><code>reverse</code>: a Boolean vector that indicates whether or
not to reverse the gene order along each gene trajectory.</p></li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_S</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddGeneBinScore.html">AddGeneBinScore</a></span><span class="op">(</span><span class="va">data_S</span>, <span class="va">gene_trajectory</span>, N.bin <span class="op">=</span> <span class="fl">5</span>, trajectories <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, assay <span class="op">=</span> <span class="st">"alra"</span>, reverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">F</span>, <span class="cn">F</span>, <span class="cn">T</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize gene bin plots for each gene trajectory</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">data_S</span>, pt.size <span class="op">=</span> <span class="fl">0.05</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Trajectory"</span>,<span class="fl">1</span>,<span class="st">"_genes"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span>, order <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://scales.r-lib.org/reference/brewer_pal.html" class="external-link">brewer_pal</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneTrajectory_files/figure-html/unnamed-chunk-10-1.png" width="1152"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">data_S</span>, pt.size <span class="op">=</span> <span class="fl">0.05</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Trajectory"</span>,<span class="fl">2</span>,<span class="st">"_genes"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span>, order <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://scales.r-lib.org/reference/brewer_pal.html" class="external-link">brewer_pal</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneTrajectory_files/figure-html/unnamed-chunk-10-2.png" width="1152"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">data_S</span>, pt.size <span class="op">=</span> <span class="fl">0.05</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Trajectory"</span>,<span class="fl">3</span>,<span class="st">"_genes"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">5</span>, order <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://scales.r-lib.org/reference/brewer_pal.html" class="external-link">brewer_pal</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoAxes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeneTrajectory_files/figure-html/unnamed-chunk-10-3.png" width="1152"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] plot3D_1.4           Matrix_1.6-0         GeneTrajectory_1.0.0</span></span>
<span><span class="co">##  [4] dplyr_1.1.3          viridis_0.6.4        viridisLite_0.4.2   </span></span>
<span><span class="co">##  [7] ggplot2_3.4.3        scales_1.2.1         SeuratObject_4.1.3  </span></span>
<span><span class="co">## [10] Seurat_4.3.0.1      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3      rstudioapi_0.15.0       jsonlite_1.8.7         </span></span>
<span><span class="co">##   [4] magrittr_2.0.3          modeltools_0.2-23       spatstat.utils_3.0-3   </span></span>
<span><span class="co">##   [7] farver_2.1.1            rmarkdown_2.24          fs_1.6.3               </span></span>
<span><span class="co">##  [10] ragg_1.2.5              vctrs_0.6.3             ROCR_1.0-11            </span></span>
<span><span class="co">##  [13] memoise_2.0.1           spatstat.explore_3.2-3  additivityTests_1.1-4.1</span></span>
<span><span class="co">##  [16] htmltools_0.5.6         sass_0.4.7              sctransform_0.3.5      </span></span>
<span><span class="co">##  [19] parallelly_1.36.0       KernSmooth_2.23-22      bslib_0.5.1            </span></span>
<span><span class="co">##  [22] htmlwidgets_1.6.2       desc_1.4.2              ica_1.0-3              </span></span>
<span><span class="co">##  [25] plyr_1.8.8              plotly_4.10.2           zoo_1.8-12             </span></span>
<span><span class="co">##  [28] cachem_1.0.8            misc3d_0.9-1            igraph_1.5.1           </span></span>
<span><span class="co">##  [31] mime_0.12               lifecycle_1.0.3         pkgconfig_2.0.3        </span></span>
<span><span class="co">##  [34] R6_2.5.1                fastmap_1.1.1           fitdistrplus_1.1-11    </span></span>
<span><span class="co">##  [37] future_1.33.0           shiny_1.7.5             digest_0.6.33          </span></span>
<span><span class="co">##  [40] rARPACK_0.11-0          colorspace_2.1-0        patchwork_1.1.3        </span></span>
<span><span class="co">##  [43] rprojroot_2.0.3         tensor_1.5              RSpectra_0.16-1        </span></span>
<span><span class="co">##  [46] irlba_2.3.5.1           textshaping_0.3.6       labeling_0.4.3         </span></span>
<span><span class="co">##  [49] progressr_0.14.0        fansi_1.0.4             spatstat.sparse_3.0-2  </span></span>
<span><span class="co">##  [52] httr_1.4.7              polyclip_1.10-4         abind_1.4-5            </span></span>
<span><span class="co">##  [55] compiler_4.3.1          withr_2.5.0             highr_0.10             </span></span>
<span><span class="co">##  [58] MASS_7.3-60             tools_4.3.1             lmtest_0.9-40          </span></span>
<span><span class="co">##  [61] flexclust_1.4-1         httpuv_1.6.11           future.apply_1.11.0    </span></span>
<span><span class="co">##  [64] goftest_1.2-3           glue_1.6.2              nlme_3.1-162           </span></span>
<span><span class="co">##  [67] promises_1.2.1          grid_4.3.1              Rtsne_0.16             </span></span>
<span><span class="co">##  [70] cluster_2.1.4           reshape2_1.4.4          generics_0.1.3         </span></span>
<span><span class="co">##  [73] gtable_0.3.4            spatstat.data_3.0-1     class_7.3-22           </span></span>
<span><span class="co">##  [76] tidyr_1.3.0             data.table_1.14.8       sp_2.0-0               </span></span>
<span><span class="co">##  [79] utf8_1.2.3              spatstat.geom_3.2-5     RcppAnnoy_0.0.21       </span></span>
<span><span class="co">##  [82] biclust_2.0.3.1         ggrepel_0.9.3           RANN_2.6.1             </span></span>
<span><span class="co">##  [85] pillar_1.9.0            stringr_1.5.0           later_1.3.1            </span></span>
<span><span class="co">##  [88] splines_4.3.1           lattice_0.21-8          FNN_1.1.3.2            </span></span>
<span><span class="co">##  [91] survival_3.5-5          deldir_1.0-9            tidyselect_1.2.0       </span></span>
<span><span class="co">##  [94] miniUI_0.1.1.1          pbapply_1.7-2           knitr_1.43             </span></span>
<span><span class="co">##  [97] gridExtra_2.3           scattermore_1.2         stats4_4.3.1           </span></span>
<span><span class="co">## [100] xfun_0.40               matrixStats_1.0.0       stringi_1.7.12         </span></span>
<span><span class="co">## [103] lazyeval_0.2.2          yaml_2.3.7              evaluate_0.21          </span></span>
<span><span class="co">## [106] codetools_0.2-19        tcltk_4.3.1             tibble_3.2.1           </span></span>
<span><span class="co">## [109] cli_3.6.1               uwot_0.1.16             xtable_1.8-4           </span></span>
<span><span class="co">## [112] reticulate_1.35.0       systemfonts_1.0.4       munsell_0.5.0          </span></span>
<span><span class="co">## [115] jquerylib_0.1.4         Rcpp_1.0.11             globals_0.16.2         </span></span>
<span><span class="co">## [118] spatstat.random_3.1-6   png_0.1-8               parallel_4.3.1         </span></span>
<span><span class="co">## [121] ellipsis_0.3.2          pkgdown_2.0.7           listenv_0.9.0          </span></span>
<span><span class="co">## [124] ggridges_0.5.4          leiden_0.4.3            purrr_1.0.2            </span></span>
<span><span class="co">## [127] rlang_1.1.1             cowplot_1.1.1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Rihao Qu, Francesco Strino.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
